File: 13/solution.py
# Path: 13/solution.py

import json


def compare(a, b):
    if type(a) == int and type(b) == int:
        return a - b

    elif type(a) == list and type(b) == list:
        for y1, y2 in zip(a, b):
            n = compare(y1, y2)

            if n != 0:
                return n

        return len(a) - len(b)

    elif type(a) == int:
        return compare([a], b)

    else:
        return compare(a, [b])


def part1(input: str) -> int:
    parsed = list(map(str.splitlines, input.split("\n\n")))

    total = 0

    for i, (a, b) in enumerate(parsed):
        x = compare(eval(a), eval(b))
        if x < 0:
            total += i + 1

    return total


def part2(input: str) -> int:
    s = [[6]]
    t = [[2]]

    parsed = list(map(eval, input.split()))
    parsed.extend([s, t])

    parsed.sort(key=lambda x: compare(x, s))
    six_location = parsed.index(s) + 1
    parsed.sort(key=lambda x: compare(x, t))
    two_location = parsed.index(t) + 1

    return six_location * two_location


if __name__ == "__main__":
    input = open("./13/input.txt", "r").read()
    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

File: 13/solution.ts
// Path: 13/solution.ts

type Input = number | Input[];

function compare(a: Input, b: Input): number {
    if (typeof a === "number" && typeof b === "number") {
        return a - b;
    } else if (Array.isArray(a) && Array.isArray(b)) {
        const zipped = a.map((e, i) => [e, b[i]]);
        for (const [aa, bb] of zipped) {
            const result = compare(aa, bb);
            if (result !== 0) {
                return result;
            }
        }
        return a.length - b.length;
    } else if (typeof a === "number") {
        return compare([a], b);
    } else if (typeof b === "number") {
        return compare(a, [b]);
    }
    return 0;
}

export function part1(input: string): number {
    const parsed = input.split("\r\n\r\n").map((e) => e.split("\r\n"));

    let total = 0;

    for (let i = 0; i < parsed.length; i++) {
        const [a, b] = parsed[i];
        const x = compare(eval(a), eval(b));
        if (x < 0) {
            total += i + 1;
        }
    }
    return total;
}

export function part2(input: string): number {
    const parsed = input.split("\r\n").map(eval);
    parsed.push([[[6]]], [[[2]]]);

    parsed.sort(compare);

    const two = parsed.findIndex((e) => compare(e, [[2]]) === 0) + 1;
    const six = parsed.findIndex((e) => compare(e, [[6]]) === 0) + 1;
    return two * six;
}

const input = Deno.readTextFileSync("./13/input.txt");

console.log("Part 1:", part1(input));
console.log("Part 2:", part2(input));
